docs/proposals/unified-backend-architecture/tech-spec.md

```diff
@@ -0,0 +1,226 @@
+
+## Multi-Tenant Architecture
+
+Mobile clients will communicate with a clusters of nodes, so the basic requirement is an ability of one node to work
+with particular number of clients simultaneously. For example, if 1 cluster has to serve 10000 mobile clients and
+this cluster consists of 100 nodes, each node has to be able to work with 100 clients simultaneously and reliably.
+Thus, the longest procedure is a new payment (creating new transaction), and even if all these 100 clients will
+sign their transactions and send it for publishing at the same time, the node must be able to handle these
+transactions reliably and as fast as possible.
+
+To achieve this goal the node must implement robust concurrent model. More specifically, all requests to the wallet
+web API endpoints must be handled concurrently.
+
+**TODO**: To think about the slowest endpoints, like `/api/v1/transactions`: is it possible to handle them
+fully concurrently? I mean publishing to the network, storing in the `wallet-db`, etc.
+
+## Scalability
+
+Initially clusters of nodes will be hosted by IOHK (on AWS EC2). And we need a mechanism of auto-scaling of the
+cluster: it must be able to grow proportionally to growing number of clients.
+
+**TODO**: To think about auto-scale supervisor. Suppose we have 1 cluster that consists of 100 nodes and each of node
+can reliably serve 100 clients. If supervisor sees that number of clients connected to this cluster is growing (and
+becomes more than 10000), it launches additional nodes.
+
+_Domen said: We might not base our autoscaling on number of users, but another resource. It's hard to say how it
+would scale before we run some benchmarks. It could be that bottleneck is CPU and you can't correlate that with number
+of subscribers._
+
+**TODO 2**: Launching additional node is relatively expensive action (in terms of time and memory). So it is better
+if one single node will be able to serve more clients than just increase a number of nodes in the cluster.
+
+## Workflow
+
+These are typical workflows, with description of API calls.
+
+### Initial State
+
+After installation of mobile client application user does not see any wallets, and he must create at least one to
+work with a client application.
+
+During the first connection with the node the new **session** is establishing, please read an explanation below.
+
+### Wallet Creation
+
+User creates a new wallet. It produces a `POST`-request to `/api/v1/wallets` (with information about client's public
+key), node creates a new wallet in `wallet-db` and sends response with information about new wallet.
+
+Even if node `N` which serves the client `C` will die, another node `M` (the client `C` will reconnect to) will be able
+to handle information about `C`'s wallet(s), because `wallet-db` is storing in the cluster.
+
+### Wallet Import/Export
+
+Wallet can be imported using client's secret key.
+
+**TODO**: To clarify this process. User can be able to obtain its secret key.
```

Comment:
Who can help us here, @vincenthz maybe? I guess this should become a little clearer once we finalise the Authentication scheme.

Reply:
Secret keys remain only on the client, so import is a frontend only operation (just turning the multi-word passphrase into the root keypair).